README.md
# Sticky Mitten Avatar API

#### This repo has been deprecated. See the  [Magnebot API](https://github.com/alters-mit/magnebot) for the updated API.

***

A high-level API for [TDW's](https://github.com/threedworld-mit/tdw/) [Sticky Mitten Avatar](https://github.com/threedworld-mit/tdw/blob/master/Documentation/misc_frontend/sticky_mitten_avatar.md). Try to gather objects and put them at a target goal area as fast as possible!

<img src="social.jpg" style="zoom:50%;" />

- Generate indoor environments with realistic rigidbody physics and visual rendering that are populated by objects. Some objects are procedurally placed, others are not.
- Use the Sticky Mitten Avatar to interact with objects. The Sticky Mitten Avatar is driven entirely by physics in a robotics-esque fashion. Its mittens can attach to objects in the scene. This API adds high-level functionality to the Sticky Mitten Avatar, such as a physics-driven IK system.
- The Sticky Mitten Avatar can place objects in containers and then pick up the container to efficiently gather objects in the scene.

## Installation

1. [Install TDW](https://github.com/threedworld-mit/tdw/). Make sure you are using v1.7.16
2. Clone this repo.
3. `cd path/to/sticky_mitten_avatar` (replace `path/to` with the actual path).
4. Install the local `sticky_mitten_avatar` pip module:

| Windows                    | OS X and Linux      |
| -------------------------- | ------------------- |
| `pip3 install -e . --user` | `pip3 install -e .` |

## Usage

```python
from sticky_mitten_avatar import StickyMittenAvatarController, Arm

c = StickyMittenAvatarController()
c.init_scene() # Creates an empty room. See API documentation for how to load a furnished scene.
task_status = c.reach_for_target(arm=Arm.left, target={"x": 0.1, "y": 0.6, "z": 0.4})
print(task_status) # TaskStatus.success
c.end()
```

## API

**[Read this document.](Documentation/sma_controller.md)** 

- At the start of the simulation, the controller caches [static object info](Documentation/static_object_info.md) per object in the scene and [static avatar info](Documentation/body_part_static.md).
- Each API function returns a [TaskStatus](Documentation/task_status.md) indicating whether the action succeeded and if not, why.
- After calling each function, the controller updates its [FrameData](Documentation/frame_data.md). This data can be used to decide what the avatar's next action will be (pick up an object, navigate around a room, etc.)
- Most complex tasks, such as navigation/pathfinding are not implemented in this API because the problem is too unbounded for a simple algorithm. However, given the output data (static object info, static avatar info, and FrameData), an agent equipped with ML data can be trained to do any of these tasks successfully.

#### API (low-level)

- For more information regarding TDW, see the [TDW repo](https://github.com/threedworld-mit/tdw/).
- For more information regarding TDW's low-level Sticky Mitten Avatar API, [read this](https://github.com/threedworld-mit/tdw/blob/master/Documentation/misc_frontend/sticky_mitten_avatar.md).

## How to write your controller

You can write your controller like this:

```python
from sticky_mitten_avatar import StickyMittenAvatarController, Arm

c = StickyMittenAvatarController()
c.init_scene()
task_status = c.reach_for_target(arm=Arm.left, target={"x": 0.1, "y": 0.6, "z": 0.4})
print(task_status) # TaskStatus.success
c.end()
```

...or this, with your own class:

```python
from sticky_mitten_avatar import StickyMittenAvatarController, Arm

class MyController(StickyMittenAvatarController):
    def my_function(self):
        return self.reach_for_target(arm=Arm.left, target={"x": 0.1, "y": 0.6, "z": 0.4})

if __name__ == "__main__":
    c = MyController()
    c.init_scene()
    print(c.my_function()) # TaskStatus.success
    c.end()
```

## Examples

All example controllers can be found in: `controllers/`

| Controller            | Description                                            |
| --------------------- | ------------------------------------------------------ |
| `fill_container.py`   | Fill a container with objects.                         |
| `social_image.py`     | Generate the social image for the GitHub preview card. |
| `initialize_scene.py` | Initialize a floorplan populated with objects.         |

## Tests

| Controller                 | Description                                                  |
| -------------------------- | ------------------------------------------------------------ |
| `ik_test.py`               | Test the IK chains of the avatar's arms.                     |
| `collision_test.py`        | Test the avatar's response to collisions.                    |
| `turn_test.py`             | Test avatar turning.                                         |
| `composite_object_test.py` | Test if the avatar can grasp a sub-object of a composite object. |
| `mitten_collision_test.py` | Test mitten collision detection.                             |
| `precision_test.py`        | Test how the `precision` parameter affects arm articulation. |
| `proc_gen_spawn_test.py`   | Test whether target objects and containers tend to stay in their initial positions. |
| `target_object_test.py`    | Test target objects.                                         |

## Utility scripts

Utility scripts are located in `util/`

| Script                      | Description                                                  |
| --------------------------- | ------------------------------------------------------------ |
| `composite_object_audio.py` | Get default audio parameters for sub-objects of composite objects. |
| `occupancy_mapper.py` | Create occupancy maps of each floorplan and layout. |
| `spawn_mapper.py` | Pre-calculate avatar spawn positions per room, per layout, per scene. |
| `room_positions.py` | Cache which positions of each occupancy map are in each room. |
| `occupancy_images.py`      | Create an image of each occupancy map per scene per layout.  |

## Troubleshooting and debugging

### "I got an error"

- Update TDW (`pip3 install tdw -U`)
- Update the TDW build to match the TDW version.
- [Read this.](https://github.com/threedworld-mit/tdw/blob/master/Documentation/misc_frontend/debug_tdw.md)
- If the stacktrace appears to lead to an error not handled in this repo or the `tdw` repo, create a GitHub issue and we'll address it as soon as possible.
- If you see this error, your Internet connection is probably not working:
```
line 35, in get_major_release
    return v.split(".")[1].strip()
IndexError: list index out of range
```

### "I can't launch in the simulation in a Docker container"

- Use the [TDW Docker image](https://github.com/threedworld-mit/tdw/blob/master/Documentation/Docker/docker.md).
- Set `launch_build=False` in the constructor; otherwise, the controller will try to launch a build on the same machine (outside of the container):

```python
from sticky_mitten_avatar import StickyMittenAvatarController

c = StickyMittenAvatarController(launch_build=False)
```

### "Images are grainy / very dark / obviously glitchy"

- [Check the player log for Open GL errors](https://github.com/threedworld-mit/tdw/blob/master/Documentation/misc_frontend/debug_tdw.md).

### "I can only initialize a blank white room"

To initialize a realistic indoor environment populated with objects, explicitly set the `scene` and `layout` parameters of [`init_scene()`](Documentation/sma_controller.md#init_scene).

```python
from sticky_mitten_avatar import StickyMittenAvatarController

c = StickyMittenAvatarController(launch_build=False)
c.init_scene(scene="1a", layout=0)
```

### "The simulation is too slow"

- Make sure you're using a GPU.
- Make sure the `demo` parameter in the constructor is `False` (the default value) and that you don't have an overhead camera. Both will render far more frames than you'll need in an actual simulation; rendering is by far the biggest factor in framerate.
- [Check the player log for Open GL errors](https://github.com/threedworld-mit/tdw/blob/master/Documentation/misc_frontend/debug_tdw.md).

### "Some actions are slow"

- Some actions will take longer to complete than others. For example, `move_forward_by(2)` will take more time to complete than `move_forward_by(1)` because the distance is longer.
- Sometimes, the avatar's interaction with objects and the environment will cause the action to take a long time. For example, if the arm of the avatar catches on a chair while the avatar is trying to move forward, it will move slower because of the chair's additional mass, friction, etc.

### "Sometimes a task fails unexpectedly / The avatar's movements are inaccurate"

This simulation is 100% physics-driven. *Every task will sometimes fail.* Possible reasons for failure include:

- The avatar tried to grasp an object but there was an obstacle in the way.
- The avatar tried to move forward but got caught on furniture.
- The avatar tried to put an object in a container but the object bounced out.

*You* will need to develop solutions to handle cases like this. You can use the [`TaskStatus`](Documentation/task_status.md) return values to figure out why a task failed and [`FrameData`](Documentation/frame_data.md) to get the current state of the simulation.

### "Objects are flying away / The avatar's mitten clipped through a held object / Physics is acting weird in general"

Physics glitches should be very rare but when they do occur, they're usually problems with the Unity physics engine that we can mitigate but not necessarily fix. If you encounter a severe glitch (such as objects falling through the floor), please create a GitHub issue and attach your controller code if possible.

### "The simulation behaves differently on different machines / Physics aren't deterministic"

We can't fix this because this is how the Unity physics engine works.

### "Objects are dragging on the ground / The avatar loses its grip on objects"

Usually this is because you called `grasp_object()` without an additional `reach_for_target()` call. `grasp_object()` will grasp the object but it won't actually lift it up.

The avatar will lose its grip on objects if it encounters a sufficiently powerful force, such as swinging the object into a wall or dragging it on the floor.

### "I need to know what each scene and layout looks like"

Images of each scene+layout combination are [here](https://github.com/alters-mit/sticky_mitten_avatar/tree/master/Documentation/images/floorplans).

Occupancy map images are [here](https://github.com/alters-mit/sticky_mitten_avatar/tree/master/images/occupancy_maps).

### "I can't navigate through the scene"

There is no built-in navigation system in this API. You'll have to write one yourself.

### "The avatar can't go to all positions on the occupancy map"

The occupancy map isn't a navigation map; some "free" positions aren't navigable.

### "I have a problem not listed here"

Create a GitHub Issue on this repo. Describe the problem and include steps to reproduce the bug. Please include your controller code if possible.

## Changelog

See [Changelog](changelog.md).

changelog.md
# Changelog

## 0.9.0

- **Deprecated this repo. See: https://github.com/alters-mit/magnebot**
- Fixed depth maps.

## 0.8.5

### Frontend

- `StickyMittenAvatarController`:
  - Renamed option `avatars` in `add_overhead_camera()` to `avatar`.
  - Fixed: Exception if the `target_object` parameter of `add_overhead_camera()` is set to None.
  - Fixed: `move_forward_by()` doesn't move backwards if `distance < 0`.
  - Fixed: `add_overhead_camera()` always disables the avatar's camera.
- `FrameData`:
  - Fixed: If there is an overhead camera in the scene, the image passes might be those rendered by the overhead camera (now, they are always the passes rendered by the avatar).

## 0.8.4

### Frontend

- `StickyMittenAvatarController`:
  - Added optional parameter `stop_on_collision` to `turn_to()`
  - Added optional parameter `stop_on_collision` to `turn_by()`
  - Parameter `stop_on_collision` in `go_to()` is used for turning as well as movement (previously, was only for movement).
- `FrameData`:
  - Fixed: Crash during an sub-action due to non-existant camera matrices data.
- Documentation:
  - Fixed: Various code examples in the API documentation have errors or are incomplete code snippets. Code examples now have correct syntax, variable names, etc. and can be run without needing to add additional lines of code.

### Backend

- Removed unused code for rolling wrist.
- Fixed: `tests/collision_test.py` doesn't work.
- Fixed: `tests/composite_object_test.py` doesn't work.
- Fixed: `tests/ik_unit_tests.py` doesn't work.

## 0.8.3

### Frontend

- `StickyMittenAvatarController`:
  - Minor speed improvement.
  - Added optional parameter `precision` to `grasp_object()`
  - Added optional parameter `precision` to `reset_arm()`
  - Added optional parameter `precision` to `drop()`
  - Added optional parameter `sub_action` to `reach_for_target()`
  - Added optional parameter `sub_action` to `drop()`
  - Added optional parameter `sub_action` to `turn_to()` 
  - Removed parameter `enable_sensor_on_finish` from `turn_to()` (replaced with `sub_action`)
  - Fixed: `drop()` returns `success` when the arm doesn't finish resetting.
- Documentation:
  
  - Improved the code examples in the API documentation and fixed some typos.
  
  - Added links in the documentation to occupancy map images and floorplan images.
  
    - Converted occupancy map images from png to jpg
    - Added floorplan images: `Documentation/images/floorplans`

### Backend

- Added optional command line arg `--floorplan` to `occupancy_images.py` to generate floorplan images instead of occupancy images.

## 0.8.2

### Frontend

- `StickyMittenAvatarController`:
  - Fixed: `rotate_camera_by()`, `reset_camera_rotation()` and `add_overhead_camera()` advance the simulation an extra frame.
  - Fixed: Sometimes, `put_in_container()` never finishes.
- Added descriptions to each category in the`StickyMittenAvatarController`  documentation.
- Added a section in the README about why actions might sometimes take a long time to complete.

## 0.8.1

### Frontend

- `StickyMittenAvatarController`:
  - Fixed: Chairs are included as valid goal positions.
- Documentation:
  - `StickyMittenAvatarController`
    - Added category headers.
    - Clarified the meaning of `init_scene()` parameters.

## 0.8.0

### Frontend

- `StickyMittenAvatarController`:
  - Removed: `pour_out_container()`
  - Removed: `shake()`
  - Removed parameter `do_motion` from `reach_for_target()`
  - Removed parameter `do_motion` from `grasp_object()`
  - Removed parameter `do_motion` from `drop()`
  - Removed parameter `do_motion` from `reset_arm()`
  - Fixed: The avatar will sometimes veer off-course when trying to move forward.
  - Fixed: The avatar will sometimes twist when bending its arms.
  - Fixed: Crash in `put_in_container()` due to the controller sometimes trying to teleport a body part of the avatar.
- `FrameData`:
  - Fixed: Crash in `get_pil_images()` if any of the image passes are None.
- `TaskStatus`:
  - Removed unused values.
- Moved the `Arm` enum class to its own file and added documentation.
- Renamed `fill_and_pour.py` to `fill_container.py` and removed the obsolete pour API call.
- Fixed a bug in `put_in_container_test.py` in which the avatar would sometimes try to pick up an object already in a container.
- Documentation:
  - README:
    - Added an intro section.
    - Added a troubleshooting section.
  - `StickyMittenAvatarController`:
    - Organized all functions by category.
    - Fixed various typos and inaccuracies.
    - Re-wrote documentation for `communicate()`.
    - Added a section describing the `Arm` type.
  - Aesthetic improvements to the headers and section breaks of each API document.

### Backend

- Added: `util/api_categories.json` Used for organizing function calls into categories in the documentation.
- Added code to `doc_gen.py` to format documentation better and organize function calls by category.
- Moved most of the functions in `util.py` to `TDWUtils`.
- Removed unused `_get_velocities()` backend function from `FrameData`.

## 0.7.0

### Frontend

- `StickyMittenAvatarController`:
  - Added: `pour_out_container()` Pour out the contents of a container.
  -  Added: optional parameter `precision` to `reach_for_target()` to adjust the threshold at which the action is considered successful.
  -  Added: optional parameter `num_attempts` to `put_in_container()` to adjust the number of attempts it makes before dropping the object.
  -  Added: optional parameter `absolute` to `reach_for_target()` for absolute world coordinates.
  -  Added: optional parameter `enable_sensor_on_finish` to `turn_to()`.
  -  Removed: parameter `num_attempts` from `put_in_container()`.
  -  Adjusted the scale of all containers, the mass of all containers, and the mass of all target objects.
  -  Added two more types of target objects.
  -  Target objects now have a random visual material.
  -  Removed `audio` parameter from constructor.

put_object_on_table.py
import numpy as np
from distutils import dir_util
from pathlib import Path
from typing import Tuple, List, Optional, Union, Dict
from tdw.controller import Controller
from tdw.tdw_utils import TDWUtils
from tdw.output_data import OutputData, AvatarChildrenNames, AvatarStickyMitten, Images, Bounds


class PutObjectOnTable(Controller):
    """
    1. Create a Sticky Mitten Avatar, a 3rd-person camera, an object, and a table.
    2. The avatar picks up the object.
    3. The avatar moves to the table.
    4. The avatar puts the object on the table.
    """

    OUTPUT_DIR = Path("images")
    # Remove the previous images.
    if OUTPUT_DIR.exists():
        dir_util.remove_tree(str(OUTPUT_DIR.resolve()))
    OUTPUT_DIR.mkdir(parents=True)
    OUTPUT_DIR = str(OUTPUT_DIR.resolve())

    def __init__(self, check_version: bool = True, launch_build: bool = True):
        print(f"Images will be saved to: {PutObjectOnTable.OUTPUT_DIR}")
        super().__init__(check_version=check_version, launch_build=launch_build)

        self.avatar_position: Tuple[float, float, float] = (0, 0, 0)
        self.frame_number = 0
        # Camera ID.
        self.cam_id = "c"

    def run(self) -> None:
        self.start()

        # Get IDs for the object, the table, and the avatars.
        o_id = self.get_unique_id()
        table_id = self.get_unique_id()
        table_position = {"x": 0, "y": 0, "z": 6.8}
        a = "a"

        # 1. Create the room.
        commands = [TDWUtils.create_empty_room(20, 20)]
        # 2. Add the avatar.
        commands.extend(TDWUtils.create_avatar(avatar_type="A_StickyMitten_Adult",
                                               avatar_id="a"))
        # 3. Add the objects. Scale the objects. Set a high mass for the table.
        # 4. Disable the avatar's cameras.
        # 5. Set the stickiness of the avatar's left mitten. Set a high drag. Rotate the head.
        # 6. Request AvatarChildrenNames data and Bounds data (this frame only).
        # 7. Strengthen the left shoulder a bit.
        commands.extend([self.get_add_object("jug05",
                                             position={"x": -0.417, "y": 0.197, "z": 0.139},
                                             rotation={"x": 90, "y": 0, "z": 0},
                                             object_id=o_id),
                         {"$type": "scale_object",
                          "scale_factor": {"x": 2, "y": 2, "z": 2},
                          "id": o_id},
                         self.get_add_object("small_table_green_marble",
                                             position={"x": 0, "y": 0, "z": 6.8},
                                             rotation={"x": 0, "y": 0, "z": 0},
                                             object_id=table_id),
                         {"$type": "scale_object",
                          "scale_factor": {"x": 2, "y": 0.5, "z": 2},
                          "id": table_id},
                         {"$type": "set_object_collision_detection_mode",
                          "id": table_id,
                          "mode": "continuous_dynamic"},
                         {"$type": "set_object_collision_detection_mode",
                          "id": o_id,
                          "mode": "continuous_dynamic"},
                         {"$type": "set_mass",
                          "mass": 100,
                          "id": table_id},
                         {"$type": "toggle_image_sensor",
                          "sensor_name": "SensorContainer",
                          "avatar_id": a},
                         {"$type": "toggle_image_sensor",
                          "sensor_name": "FollowCamera",
                          "avatar_id": a},
                         {"$type": "set_stickiness",
                          "sub_mitten": "back",
                          "sticky": True,
                          "is_left": True,
                          "avatar_id": a},
                         {"$type": "set_avatar_drag",
                          "drag": 1000,
                          "angular_drag": 1000,
                          "avatar_id": a},
                         {"$type": "rotate_head_by",
                          "axis": "pitch",
                          "angle": 5},
                         {"$type": "send_avatar_children_names",
                          "ids": [a],
                          "frequency": "once"},
                         {"$type": "send_bounds",
                          "ids": [table_id],
                          "frequency": "once"},
                         {"$type": "adjust_joint_force_by",
                          "delta": 2,
                          "joint": "shoulder_left",
                          "axis": "pitch"}])
        # 7. Add a 3rd-person camera.
        commands.extend(TDWUtils.create_avatar(avatar_type="A_Img_Caps_Kinematic",
                                               avatar_id="c",
                                               position={"x": -3.9, "y": 2.3, "z": 4.3}))
        # 8. Request StickyMittenAvatar and Images data per-frame.
        commands.extend([{"$type": "set_pass_masks",
                          "pass_masks": ["_img"],
                          "avatar_id": self.cam_id},
                         {"$type": "send_avatars",
                          "ids": [a],
                          "frequency": "always"},
                         {"$type": "send_images",
                          "ids": [self.cam_id],
                          "frequency": "always"}])

        resp = self.communicate(commands)
        # Get the object ID of the left mitten, the size of the table, and the avatar's position.
        mitten_left_id = None
        table_size: Tuple[float, float, float] = (0, 0, 0)

        for r in resp[:-1]:
            r_id = OutputData.get_data_type_id(r)
            # Get the mitten ID.
            if r_id == "avcn":
                avcn = AvatarChildrenNames(r)
                for i in range(avcn.get_num_children()):
                    if avcn.get_child_name(i) == "mitten_left":
                        mitten_left_id = avcn.get_child_id(i)
            # Get the table bounds.
            elif r_id == "boun":
                boun = Bounds(r)
                for i in range(boun.get_num()):
                    if boun.get_id(i) == table_id:
                        table_size = (boun.get_right(i)[0] - boun.get_left(i)[0],
                                      boun.get_top(i)[1] - boun.get_bottom(i)[1],
                                      boun.get_front(i)[2] - boun.get_back(i)[2])
            # Get the avatar's position.
            elif r_id == "avsm":
                self.avatar_position = AvatarStickyMitten(r).get_position()

        # Pick up the object.
        # Lift up the object.
        self._bend_arm_joints([{"$type": "pick_up_proximity",
                                "distance": 20,
                                "grip": 10000,
                                "is_left": True,
                                "avatar_id": a},
                               {"$type": "rotate_head_by",
                                "axis": "pitch",
                                "angle": 20,
                                "avatar_id": a},
                               {"$type": "bend_arm_joint_by",
                                "angle": 25,
                                "joint": "shoulder_left",
                                "axis": "pitch",
                                "avatar_id": a},
                               {"$type": "bend_arm_joint_by",
                                "angle": -25,
                                "joint": "shoulder_left",
                                "axis": "yaw",
                                "avatar_id": a},
                               {"$type": "bend_arm_joint_by",
                                "angle": 60,
                                "joint": "shoulder_left",
                                "axis": "roll",
                                "avatar_id": a},
                               {"$type": "bend_arm_joint_by",
                                "angle": 100,
                                "joint": "elbow_left",
                                "axis": "pitch",
                                "avatar_id": a}])
        # Allow the avatar to move again.
        self._do_frame({"$type": "set_avatar_drag",
                        "drag": 0.125,
                        "angular_drag": 1000,
                        "avatar_id": a})

        # Move to the table.
        move_to_table = True
        # The position of the side of the table the avatar is aiming for.
        table_side_position = {"x": table_position["x"], "y": 0, "z": table_position["z"] - table_size[2]}
        while move_to_table:
            # Stop moving if we are close enough.
            if TDWUtils.get_distance(table_side_position, TDWUtils.array_to_vector3(self.avatar_position)) < 0.7:
                move_to_table = False
            # Keep moving forward.
            else:
                self._do_frame({"$type": "move_avatar_forward_by",
                                "avatar_id": a,
                                "magnitude": 20})
        # Stop.
        # Allow the avatar to move again.
        self._do_frame({"$type": "set_avatar_drag",
                        "drag": 1000,
                        "angular_drag": 1000,
                        "avatar_id": a})

        mitten_over_table = False
        target_z = table_side_position["z"] + 0.17
        # Keep lifting until the mitten is over the table.
        while not mitten_over_table:
            avsm = self._bend_arm_joints([{"$type": "bend_arm_joint_by",
                                           "angle": 15,
                                           "joint": "shoulder_left",
                                           "axis": "pitch",
                                           "avatar_id": a},
                                          {"$type": "bend_arm_joint_by",
                                           "angle": -10,
                                           "joint": "elbow_left",
                                           "axis": "pitch",
                                           "avatar_id": a}])
            # Get the mitten.
            for i in range(avsm.get_num_body_parts()):
                if avsm.get_body_part_id(i) == mitten_left_id:
                    # Check if the mitten is over the table.
                    mitten_over_table = avsm.get_body_part_position(i)[2] >= target_z

        # Drop the arm gently.
        mitten_near_table = False
        while not mitten_near_table:
            avsm = self._bend_arm_joints([{"$type": "bend_arm_joint_by",
                                           "angle": -25,
                                           "joint": "shoulder_left",
                                           "axis": "pitch",
                                           "avatar_id": a}])
            # Get the mitten.
            for i in range(avsm.get_num_body_parts()):
                if avsm.get_body_part_id(i) == mitten_left_id:
                    # Check if the mitten very close to the table surface.
                    mitten_near_table = avsm.get_body_part_position(i)[1] - table_size[1] <= 0.1
        # Drop the object.
        # Allow the avatar to move again.
        self._do_frame([{"$type": "put_down",
                         "is_left": True,
                        "avatar_id": a},
                        {"$type": "set_avatar_drag",
                         "drag": 0.125,
                         "angular_drag": 1000,
                         "avatar_id": a}])
        # Back away from the table.
        mitten_away_from_table = False
        while not mitten_away_from_table:
            avsm = self._do_frame({"$type": "move_avatar_forward_by",
                                   "avatar_id": a,
                                   "magnitude": -50})
            # Get the mitten.
            for i in range(avsm.get_num_body_parts()):
                if avsm.get_body_part_id(i) == mitten_left_id:
                    # Check if the mitten has cleared the table.
                    mitten_away_from_table = avsm.get_body_part_position(i)[2] < target_z
        # Drop the arm.
        self._do_frame([{"$type": "rotate_head_by",
                         "axis": "pitch",
                         "angle": -25},
                        {"$type": "bend_arm_joint_to",
                         "angle": 0,
                         "joint": "shoulder_left",
                         "axis": "pitch",
                         "avatar_id": a},
                        {"$type": "bend_arm_joint_to",
                         "angle": 0,
                         "joint": "shoulder_left",
                         "axis": "yaw",
                         "avatar_id": a},
                        {"$type": "bend_arm_joint_to",
                         "angle": 0,
                         "joint": "shoulder_left",
                         "axis": "roll",
                         "avatar_id": a},
                        {"$type": "bend_arm_joint_to",
                         "angle": 0,
                         "joint": "elbow_left",
                         "axis": "pitch",
                         "avatar_id": a}])
        # Back away from the table.
        away_from_table = False
        while not away_from_table:
            self._do_frame({"$type": "move_avatar_forward_by",
                            "avatar_id": a,
                            "magnitude": -20})
            away_from_table = TDWUtils.get_distance(TDWUtils.array_to_vector3(self.avatar_position),
                                                    table_position) > 1.8
        # Let the joints drop.
        self._bend_arm_joints([])
        self.communicate({"$type": "terminate"})

    def _do_frame(self, commands: Union[List[dict], dict]) -> AvatarStickyMitten:
        """
        Send commands to the build. Receive images and sticky mitten avatar data.
        Save the images.

        :param commands: The commands for this frame.

        :return: The avatar data.
        """

        # Add to the list of commands: Look at the avatar.
        frame_commands = [{"$type": "look_at_position",
                           "position": {"x": self.avatar_position[0],
                                        "y": self.avatar_position[1] + 0.5,
                                        "z": self.avatar_position[2]},
                           "avatar_id": self.cam_id}]

        if isinstance(commands, dict):
            frame_commands.append(commands)
        else:
            frame_commands.extend(commands)

        resp = self.communicate(frame_commands)

        avsm: Optional[AvatarStickyMitten] = None
        for r in resp[:-1]:
            r_id = OutputData.get_data_type_id(r)
            # Save images.
            if r_id == "imag":
                TDWUtils.save_images(images=Images(r),
                                     filename=TDWUtils.zero_padding(self.frame_number, 4),
                                     output_directory=PutObjectOnTable.OUTPUT_DIR)
            elif r_id == "avsm":
                avsm = AvatarStickyMitten(r)
        # Update the position of the avatar and the frame number.
        self.avatar_position = avsm.get_position()
        self.frame_number += 1
        return avsm

    def _bend_arm_joints(self, commands: Union[List[dict], dict]) -> AvatarStickyMitten:
        """
        Send commands to bend the arm joints. Wait until the joints stop moving.

        :param commands: The commands for this frame.

        :return: The avatar data.
        """

        avsm = self._do_frame(commands)

        # Get the body part rotations.
        body_part_rotations: Dict[int, np.array] = dict()
        for i in range(avsm.get_num_body_parts()):
            body_part_rotations[avsm.get_body_part_id(i)] = np.array(avsm.get_body_part_rotation(i))
        # Wait for the joints to stop rotating.
        done_rotating = False
        while not done_rotating:
            avsm = self._do_frame([])
            bpr: Dict[int, np.array] = dict()
            done_rotating = True
            for j in range(avsm.get_num_body_parts()):
                b_id = avsm.get_body_part_id(j)
                bpr[b_id] = np.array(avsm.get_body_part_rotation(j))
                if np.linalg.norm(bpr[b_id] - body_part_rotations[b_id]) > 0.01:
                    done_rotating = False
            body_part_rotations.clear()
            for b_id in bpr:
                body_part_rotations[b_id] = bpr[b_id]
        return avsm


if __name__ == "__main__":
    PutObjectOnTable().run()
